source/wrapper/contracts/Wrapper.sol

```diff
@@ -74,42 +75,65 @@ contract Wrapper {
     require(order.signature.v != 0,
       "SIGNATURE_MUST_BE_SENT");
 
-    // The sender is sending ether that must be wrapped.
-    if (order.sender.token == address(wethContract)) {
+    // Check if sender is sending ether that must be wrapped.
+    _wrapEther(order.sender);
 
-      // Ensure message value is sender param.
-      require(order.sender.param == msg.value,
+    // Perform the swap.
+    swapContract.swap(order);
+
+    // Check if sender is receiving ether that must be unwrapped.
+    _unwrapEther(order.sender.wallet, order.signer.token, order.signer.param);
+  }
+
+  function provideDelegateOrder(
+    Types.Order calldata order,
+    IDelegate delegate
+  ) external payable {
+    // Delegate trade wallet must be order sender - checked in Delegate
+    // Signature must be sent - checked in Delegate
+
+    // Check if signer is sending ether that must be wrapped.
+    _wrapEther(order.signer);
+
+    // Provide the order to the Delegate.
+    delegate.provideOrder(order);
+
+    // Check if signer is receiving ether that must be unwrapped.
+    _unwrapEther(order.signer.wallet, order.sender.token, order.sender.param);
+  }
+
+  function _wrapEther(Types.Party memory party) internal {
+    // Check whether ether needs wrapping
+    if (party.token == address(wethContract)) {
+      // Ensure message value is param.
+      require(party.param == msg.value,
         "VALUE_MUST_BE_SENT");
 
       // Wrap (deposit) the ether.
       wethContract.deposit.value(msg.value)();
 
-      // Transfer the WETH from the wrapper to sender.
-      wethContract.transfer(order.sender.wallet, order.sender.param);
+      // Transfer the WETH from the wrapper to party.
+      wethContract.transfer(party.wallet, party.param);
```

Comment:
is it worth making a note, require() or return value is not checked due to weth not returning false on failure?